mongocrypt/src/ctx.rs

```diff
@@ -0,0 +1,687 @@
+            }
+        }
+        Ok(self.into_ctx())
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[non_exhaustive]
+pub enum Algorithm {
+    AeadAes256CbcHmacSha512Deterministic,
+    AeadAes256CbcHmacSha512Random,
+}
+
+impl Algorithm {
+    fn c_str(&self) -> &'static CStr {
+        let bytes: &[u8] = match self {
+            Self::AeadAes256CbcHmacSha512Deterministic => {
+                b"AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic\0"
+            }
+            Self::AeadAes256CbcHmacSha512Random => b"AEAD_AES_256_CBC_HMAC_SHA_512-Random\0",
+        };
+        unsafe { CStr::from_bytes_with_nul_unchecked(bytes) }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[non_exhaustive]
+pub enum IndexType {
+    None,
+    Equality,
+}
+
+impl IndexType {
+    fn as_native(self) -> sys::mongocrypt_index_type_t {
+        match self {
+            IndexType::None => sys::mongocrypt_index_type_t_MONGOCRYPT_INDEX_TYPE_NONE,
+            IndexType::Equality => sys::mongocrypt_index_type_t_MONGOCRYPT_INDEX_TYPE_EQUALITY,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[non_exhaustive]
+pub enum QueryType {
+    Equality,
+}
+
+impl QueryType {
+    fn as_native(self) -> sys::mongocrypt_query_type_t {
+        match self {
+            QueryType::Equality => sys::mongocrypt_query_type_t_MONGOCRYPT_QUERY_TYPE_EQUALITY,
+        }
+    }
+}
+
+pub struct Ctx {
+    inner: OwnedPtr<sys::mongocrypt_ctx_t>,
+}
+
+impl HasStatus for Ctx {
+    unsafe fn native_status(&self, status: *mut sys::mongocrypt_status_t) {
+        sys::mongocrypt_ctx_status(*self.inner.borrow(), status);
+    }
+}
+
+/// Manages the state machine for encryption or decryption.
+impl Ctx {
+    /// Get the current state of a context.
+    pub fn state(&self) -> Result<State> {
+        let s = unsafe { sys::mongocrypt_ctx_state(*self.inner.borrow()) };
+        if s == sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_ERROR {
+            return Err(self.status().as_error());
+        }
+        Ok(State::from_native(s))
+    }
+
+    /// Get BSON necessary to run the mongo operation when in `State::NeedMongo*` states.
+    ///
+    /// The returned value:
+    /// * for `State::NeedMongoCollinfo` it is a listCollections filter.
+    /// * for `State::NeedMongoKeys` it is a find filter.
+    /// * for `State::NeedMongoMarkings` it is a command to send to mongocryptd.
+    pub fn mongo_op(&self) -> Result<&RawDocument> {
+        // Safety: `mongocrypt_ctx_mongo_op` updates the passed-in `Binary` to point to a chunk of
+        // BSON with the same lifetime as the underlying `Ctx`.  The `Binary` itself does not own
+        // the memory, and gets cleaned up at the end of the unsafe block.  Lifetime inference on
+        // the return type binds `op_bytes` to the same lifetime as `&self`, which is the correct
+        // one.
+        let op_bytes = unsafe {
+            let bin = Binary::new();
```

Comment:
won't `Binary`'s `drop` invoke `mongocrypt_binary_destroy` at the end of this block, which I think we don't want here?

Reply:
A `mongocrypt_binary_t` is a non-owning view (equivalent to a slice), so `mongocrypt_binary_destroy` doesn't deallocate the memory it points to.  The `Binary` created here is used as an out-param in `mongocrypt_ctx_mongo_op`; the [documentation for that](https://github.com/mongodb/libmongocrypt/blob/master/src/mongocrypt.h#L1040) says that the memory referenced is valid as long as the ctx is, which is equivalent to the lifetime of `&self` in that method.